sys/net/routing/ng_rpl/ng_rpl_dodag.c

```diff
@@ -0,0 +1,384 @@
+        (*dodag)->lifetime_unit = NG_RPL_LIFETIME_UNIT;
+        (*dodag)->node_status = NG_RPL_NORMAL_NODE;
+        (*dodag)->dao_seq = NG_RPL_COUNTER_INIT;
+        (*dodag)->dtsn = 0;
+        (*dodag)->dao_ack_received = false;
+        (*dodag)->dao_counter = 0;
+        (*dodag)->parents = NULL;
+        (*dodag)->cleanup_time = timex_set(NG_RPL_CLEANUP_TIME, 0);
+        return true;
+    }
+
+    /* no space available to allocate a new dodag */
+    DEBUG("Could not allocate a new RPL DODAG\n");
+    *dodag = NULL;
+    return false;
+}
+
+bool ng_rpl_dodag_remove(ng_rpl_dodag_t *dodag)
+{
+    ng_rpl_dodag_remove_all_parents(dodag);
+    ng_rpl_instance_t *inst = dodag->instance;
+    LL_DELETE(inst->dodags, dodag);
+    trickle_stop(&dodag->trickle);
+    vtimer_remove(&dodag->dao_timer);
+    vtimer_remove(&dodag->cleanup_timer);
+    memset(dodag, 0, sizeof(ng_rpl_dodag_t));
+    if (inst->dodags == NULL) {
+        ng_rpl_instance_remove(inst);
+    }
+    return true;
+}
+
+void ng_rpl_dodag_remove_all_parents(ng_rpl_dodag_t *dodag)
+{
+    ng_rpl_parent_t *elt, *tmp;
+    LL_FOREACH_SAFE(dodag->parents, elt, tmp) {
+        ng_rpl_parent_remove(elt);
+    }
+    vtimer_remove(&dodag->cleanup_timer);
+    vtimer_set_msg(&dodag->cleanup_timer, dodag->cleanup_time, ng_rpl_pid,
+            NG_RPL_MSG_TYPE_CLEANUP_HANDLE, dodag);
+    return;
+}
+
+ng_rpl_dodag_t *ng_rpl_dodag_get(ng_rpl_instance_t *instance, ng_ipv6_addr_t *dodag_id)
+{
+    if ((instance == NULL) || instance->state == 0) {
+        DEBUG("Instance is NULL or unused\n");
+        return NULL;
+    }
+
+    ng_rpl_dodag_t *dodag = NULL;
+    LL_FOREACH(instance->dodags, dodag) {
+        if (ng_ipv6_addr_equal(&dodag->dodag_id, dodag_id)) {
+            return dodag;
+        }
+    }
+    return NULL;
+}
+
+bool ng_rpl_parent_add_by_addr(ng_rpl_dodag_t *dodag, ng_ipv6_addr_t *addr, ng_rpl_parent_t **parent)
+{
+    if ((dodag == NULL) || dodag->state == 0) {
+        DEBUG("DODAG is NULL or unused\n");
+        return false;
+    }
+
+    *parent = NULL;
+    bool first = true;
+    for (uint8_t i = 0; i < NG_RPL_PARENTS_NUMOF; ++i) {
+        /* save position to the first unused instance */
+        if ((ng_rpl_parents[i].state == 0) && first) {
+            *parent = &ng_rpl_parents[i];
+            first = false;
+            continue;
+        }
+        /* return false if parent exists */
+        else if ((ng_rpl_parents[i].state != 0) &&
+                (ng_rpl_parents[i].dodag->instance->id == dodag->instance->id) &&
+                ng_ipv6_addr_equal(&ng_rpl_parents[i].dodag->dodag_id, &dodag->dodag_id) &&
+                ng_ipv6_addr_equal(&ng_rpl_parents[i].addr, addr)) {
+            DEBUG("parent with addr: %s does exist\n", ng_ipv6_addr_to_str(addr_str, addr,
+                        sizeof(addr_str)));
+            *parent = &ng_rpl_parents[i];
+            return false;
+        }
+    }
+
+    if (*parent != NULL) {
```

Comment:
the same logic applies here. `parent` is a pointer to a pointer, therefore the dereference is needed